src/sentry/utils/snuba.py

```diff
@@ -194,22 +183,93 @@ def flat_conditions(conditions):
-    for a given model and return a lookup dictionary from those keys to the
-    equivalent ones in snuba.
+    name instead of the the environment ID. Here we create and return forward()
+    and reverse() translation functions that perform all the required changes.
+
+    forward() is designed to work on the filter_keys and so should be called
+    with a map of {column: [key1, key2], ...} and should return an updated map
+    with the filter keys replaced with the ones that Snuba expects.
+
+    reverse() is designed to work on result rows, so should be called with a row
+    in the form {column: value, ...} and will return a translated result row.
+
+    Because translation can potentially rely on combinations of different parts
+    of the result row, I decided to implement them as composable functions over the
+    row to be translated. This should make it simpler to add any other needed
+    translations as long as you can express them as forward(filters) and reverse(row)
+    functions.
     """
-    mappings = {
+
+    def identity(x):
+        return x
+
+    def compose(f, g):
+        return lambda x: f(g(x))
+
+    def replace(d, key, val):
+        return d.update({key: val}) or d
+
+    forward = identity
+    reverse = identity
+
+    map_columns = {
         'environment': (Environment, 'name', lambda name: None if name == '' else name),
-        'tags[sentry:release]': (Release, 'version', lambda name: name),
+        'tags[sentry:release]': (Release, 'version', identity),
     }
-    if column in mappings and ids:
-        model, field, transform = mappings[column]
-        objects = model.objects.filter(id__in=ids).values_list('id', field)
-        return {k: transform(v) for k, v in objects}
-    return None
+
+    for col, (model, field, fmt) in six.iteritems(map_columns):
+        fwd, rev = None, None
+        ids = filter_keys.get(col)
+        if not ids:
+            continue
+        if is_grouprelease and col == 'tags[sentry:release]':
+            # GroupRelease -> Release translation is a special case because the
+            # translation relies on both the Group and Release value in the result row.
+            gr_map = GroupRelease.objects.filter(id__in=ids)\
+                                         .values_list('id', 'group_id', 'release_id')
+            versions = dict(Release.objects.filter(id__in=[x[2] for x in gr_map])
+                                           .values_list('id', 'version'))
+
+            fwd_map = dict((gr, (group, versions[release])) for (gr, group, release) in gr_map)
+            rev_map = dict(reversed(t) for t in six.iteritems(fwd_map))
+            fwd = (
+                lambda col, trans:
+                    lambda filters: replace(filters, col, [trans[k][1] for k in filters[col]])
+            )(col, fwd_map)
+            rev = (
+                lambda col, trans:
+                    lambda row: replace(row, col, trans.get((row.get('issue'), row.get(col))))
```

Comment:
Why are these `row.get(key)` instead of `row[key]`? Are there cases where you expect the key to be missing here?

Reply:
Good point, I think changing these to `get()` was a somewhat reactionary change to a test failure I had, but I'll look into it a little more deeply to see if there is a more fundamental way to deal with missing keys instead of newspapering over it here.